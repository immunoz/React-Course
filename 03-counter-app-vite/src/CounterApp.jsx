import React, { useState } from 'react'
import PropTypes from 'prop-types'

function CounterApp({ value }) {

    const [ counter, setCounter ] = useState( value );

    const increaseCounter = () => {
        setCounter( (c) => c + 1 );
    }
    
    const subtractCounter = () => {
        setCounter( (c) => c - 1 );
    }
    
    const resetCounter = () => {
        setCounter( value );
    }

    return (
        <>
            <h1>Counter App</h1>
            <h2 data-testid='counter'>{ counter }</h2>

            <button onClick={ increaseCounter }> +1 </button>
            <button onClick={ subtractCounter }> -1 </button>
            <button onClick={ resetCounter }> reset </button>

        </>
    )
}

export default CounterApp

CounterApp.defaultProps = {
    value: 0,
}

PropTypes.propTypes = {
    value: PropTypes.number,
}