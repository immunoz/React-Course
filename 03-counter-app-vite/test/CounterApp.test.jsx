import { fireEvent, render, screen } from "@testing-library/react"
import CounterApp from "../src/CounterApp";

describe('Testing the component <CounterApp />', () => {

    const defaultValueForTesting = 100;

    // IDs
    const counterNumberID = 'counter';

    test('should match snapshot', () => {

        const { container } = render( <CounterApp value={ defaultValueForTesting }/> );
        expect( container ).toMatchSnapshot();
      
    });

    test('should show the initial value 100', () => {

        render( <CounterApp value={ defaultValueForTesting }/> );
        expect( screen.getByTestId( counterNumberID ).innerHTML ).toBe('100');
      
    });

    test('should increment with the +1 button', () => {

        render( <CounterApp value={ defaultValueForTesting }/> );
        fireEvent.click( screen.getByText('+1') );

        expect( screen.getByTestId( counterNumberID ).innerHTML ).toBe('101');
      
    });

    test('should increment with the -1 button', () => {

        render( <CounterApp value={ defaultValueForTesting }/> );
        fireEvent.click( screen.getByText('-1') );

        expect( screen.getByTestId( counterNumberID ).innerHTML ).toBe('99');
      
    });

    test('should reset the counter to the default value', () => {
      
        render( <CounterApp value={ defaultValueForTesting }/> );
        fireEvent.click( screen.getByText('+1') );
        fireEvent.click( screen.getByText('+1') );
        fireEvent.click( screen.getByText('+1') );
        fireEvent.click( screen.getByText('reset') );

        expect( screen.getByTestId( counterNumberID ).innerHTML ).toBe(defaultValueForTesting.toString());

    })
    
    
});
