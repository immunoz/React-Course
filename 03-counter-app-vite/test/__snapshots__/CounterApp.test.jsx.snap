// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing the component <CounterApp /> should match snapshot 1`] = `
<div>
  <h1>
    Counter App
  </h1>
  <h2
    data-testid="counter"
  >
    100
  </h2>
  <button>
     +1 
  </button>
  <button>
     -1 
  </button>
  <button>
     reset 
  </button>
</div>
`;
